feat: support blank/omitted database in connections.json

CarbonChauvinist opened the pull request:
This emulates VSCodes behavior for connection entries where the database field can be omitted entirely, or simply left as "" in which case will default to "master".

Fixes #94

CarbonChauvinist commented:
@Kurren123 Hope you're okay with this attempt at addressing. Please let me know your thoughts once you've had some time to review.

Kurren123 commented:
Thanks for contributing! In vscode if you don’t provide the db name in the connection json, does it send “master” to the language server? Or does it send no db name and the language server default it to “master”?

I have a feeling the language server might default it to master, in case the default database name is something other than master (can it be changed?).

CarbonChauvinist commented:
@Kurren123 Hmmm, good point. I've just checked and here's the intercepted payloads from VSC connecting to a localhost server for each option (omitted or blank):

<details>
<summary> "database": "" </summary>

```
=== CLIENT -> SERVER ===
Content-Length: 423

{"jsonrpc":"2.0","id":1,"method":"connection/connect","params":{"ownerUri":"untitled:Untitled-1","connection":{"options":{"id":"8FED358D-F6DB-44A1-90F1-8F0A34DE3C2A","server":"localhost","database":"","databaseDisplayName":"","user":"ghost","password":"mysupersecretpassword","authenticationType":"SqlLogin","encrypt":"Mandatory","trustServerCertificate":true,"connectTimeout":15,"commandTimeout":30,"applicationName":"vscode-mssql"}}}}

=== SERVER -> CLIENT ===
Content-Length: 40

{"jsonrpc":"2.0","id":"1","result":true}

=== SERVER -> CLIENT ===
Content-Length: 739

{"jsonrpc":"2.0","method":"connection/complete","params":{"ownerUri":"untitled:Untitled-1","connectionId":"87ac50c9-475f-4467-86d7-97165094e423","messages":null,"errorMessage":null,"errorNumber":null,"serverInfo":{"cpuCount":8,"physicalMemoryInMB":6245,"serverMajorVersion":16,"serverMinorVersion":0,"serverReleaseVersion":4195,"engineEditionId":3,"serverVersion":"16.0.4195.2","serverLevel":"RTM","serverEdition":"Developer Edition (64-bit)","isCloud":false,"azureVersion":0,"osVersion":"Arch Linux Unknown","machineName":"bythos","options":{}},"connectionSummary":{"serverName":"localhost","databaseName":"master","userName":"ghost"},"type":"Default","isSupportedVersion":true,"unsupportedVersionMessage":null,"serverConnectionId":"55"}}
```
</details>


<details>
<summary> omitted </summary>

```
=== CLIENT -> SERVER ===
Content-Length: 384

{"jsonrpc":"2.0","id":1,"method":"connection/connect","params":{"ownerUri":"untitled:Untitled-1","connection":{"options":{"id":"8FED358D-F6DB-44A1-90F1-8F0A34DE3C2A","server":"localhost","user":"ghost","password":"mysupersecretpassword","authenticationType":"SqlLogin","encrypt":"Mandatory","trustServerCertificate":true,"connectTimeout":15,"commandTimeout":30,"applicationName":"vscode-mssql"}}}}

=== SERVER -> CLIENT ===
Content-Length: 40

{"jsonrpc":"2.0","id":"1","result":true}

=== SERVER -> CLIENT ===
Content-Length: 739

{"jsonrpc":"2.0","method":"connection/complete","params":{"ownerUri":"untitled:Untitled-1","connectionId":"a7a9c534-d2a7-432d-8565-f79f734157c1","messages":null,"errorMessage":null,"errorNumber":null,"serverInfo":{"cpuCount":8,"physicalMemoryInMB":6245,"serverMajorVersion":16,"serverMinorVersion":0,"serverReleaseVersion":4195,"engineEditionId":3,"serverVersion":"16.0.4195.2","serverLevel":"RTM","serverEdition":"Developer Edition (64-bit)","isCloud":false,"azureVersion":0,"osVersion":"Arch Linux Unknown","machineName":"bythos","options":{}},"connectionSummary":{"serverName":"localhost","databaseName":"master","userName":"ghost"},"type":"Default","isSupportedVersion":true,"unsupportedVersionMessage":null,"serverConnectionId":"56"}}
```
</details>

Unless I'm missing something the "master" magically appears from the server?

Kurren123 commented:
Okay, so it looks like we just need to send whatever is in the connections json, look at the connection complete response and set the connection details in the query manager with the response contents.

CarbonChauvinist committed:
feat: support blank/omitted database in connections.json

Mirror VSCode's behavior by passing through connections.json values
as-is and using the server's connection response to confirm the database
name when blank or omitted.

Fixes #94

CarbonChauvinist commented:
@Kurren123 Okay have made the revisions - actually seems really straight forward, feel bad it took me so long to carve out the time. Let me know your thoughts once you've had a chance to take a look.

Kurren123 commented:
Thanks for the feature!
